zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnCompensationSubscriptionBehaviour.java

```diff
@@ -109,51 +124,86 @@ private Optional<String> getCompensationHandlerId(final ExecutableActivity eleme
+      final ExecutableFlowElement element,
+      final ExecutableActivity compensationActivity,
+      final BpmnElementContext context) {
+    final String compensationActivityId = BufferUtil.bufferAsString(compensationActivity.getId());
+    // trigger the compensation for the given activity
+    return triggerCompensationInScope(
+        element,
+        context,
+        subscription ->
+            subscription.getRecord().getCompensableActivityId().equals(compensationActivityId));
+  }
+
+  private boolean triggerCompensationInScope(
+      final ExecutableFlowElement element,
+      final BpmnElementContext context,
+      final Predicate<CompensationSubscription> subscriptionFilter) {
+    final List<Long> compensationScopeKeys = getCompensationScopeKeys(element, context);
+
     // ignore subscriptions that are already triggered
-    final var subscriptions =
+    final var notTriggeredSubscriptions =
         compensationSubscriptionState
             .findSubscriptionsByProcessInstanceKey(
                 context.getTenantId(), context.getProcessInstanceKey())
             .stream()
             .filter(not(BpmnCompensationSubscriptionBehaviour::isCompensationTriggered))
             .toList();
 
-    if (hasCompensationSubscriptionInScope(subscriptions, context.getFlowScopeKey())) {
-      // trigger the compensation in the current scope and propagate to the subprocesses
-      triggerCompensationFromTopToBottom(context, subscriptions, context.getFlowScopeKey());
-      return true;
+    // filter subscriptions by their scope
+    final var subscriptionsWithinScope =
+        notTriggeredSubscriptions.stream()
+            .filter(
+                subscription ->
+                    compensationScopeKeys.contains(
+                        subscription.getRecord().getCompensableActivityScopeKey()))
+            .filter(subscriptionFilter)
+            .toList();
+
+    if (subscriptionsWithinScope.isEmpty()) {
+      return false; // no subscriptions are triggered
+    }
+
+    // trigger the compensation for all subscriptions in the scopes
+    subscriptionsWithinScope.forEach(
+        subscription ->
+            triggerCompensationForSubscription(context, notTriggeredSubscriptions, subscription));
+    return true;
+  }
+
+  private List<Long> getCompensationScopeKeys(
+      final ExecutableFlowElement element, final BpmnElementContext context) {
+    final long compensationEventScopeKey = context.getFlowScopeKey();
+
+    if (isElementInsideEventSubprocess(element)) {
+      // inside an event subprocess, trigger the compensation also from outside the event subprocess
+      final BpmnElementContext flowScopeContext = stateBehavior.getFlowScopeContext(context);
+      final long eventSubprocessScopeKey = flowScopeContext.getFlowScopeKey();
+
+      return List.of(compensationEventScopeKey, eventSubprocessScopeKey);
 
     } else {
-      return false;
+      return List.of(compensationEventScopeKey);
```

Comment:
🔧 You can refactor this method, using the exit condition first `!isElementInsideEventSubprocess(element)`

Reply:
True. In this case, I prefer having a positive condition instead of negation (i.e. `!is..`).